File: Graph/UFS/947_MaxStonestoMove.py
from typing import List
from collections import defaultdict
class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        son_father = dict()
        root_sons = defaultdict(set)

        for x, y in stones:
            y += 1<<14
            fx, fy = son_father.get(x, -1), son_father.get(y, -1)
            if fx == -1:
                son_father[x] = x
                root_sons[x] = set()
                root_sons[x].add(x)
                root_sons[x].add(y)
                
            if fy == -1:
                son_father[y] = y
                root_sons[y] = set()
                root_sons[y].add(x)
                root_sons[y].add(y)

            p1, np1 = x, son_father[x]
            p2, np2 = y, son_father[y]
            while p1 != np1:
                p1, np1 = np1, son_father[np1]
            while p2 != np2:
                p2, np2 = np2, son_father[np2]
            
            if p1!=p2:
                son_father[p2] = p1
                for p2_son in root_sons[p2]:
                    root_sons[p1].add(p2_son)
                    son_father[p2_son] = p1
                root_sons.pop(p2)
        return len(stones) - len(root_sons)

File: Graph/UFS/947_MaxStonesToMove.cpp
#include <bits/stdc++.h>

using namespace std;
class Solution {
public:
    int removeStones(vector<vector<int>> &stones) {
        map<int, int> son_father;
        map<int, set<int>> root_sons;

        for (auto &stone : stones) {
            int x = stone[0], y = stone[1] + (1 << 14);
            if (son_father.find(x) == son_father.end()) {
                son_father[x] = x;
                set<int> hh;
                hh.emplace(x);
                hh.emplace(y);
                root_sons[x] = hh;

            }
            if (son_father.find(y) == son_father.end()) {
                son_father[y] = y;
                set<int> hh;
                hh.emplace(x);
                hh.emplace(y);
                root_sons[y] = hh;
            }
            int rt1 = x, nrt1 = son_father[x],
                rt2 = y, nrt2 = son_father[y];
            while (rt1 != nrt1) {
                rt1 = nrt1;
                nrt1 = son_father[nrt1];
            }
            while (rt2 != nrt2) {
                rt2 = nrt2;
                nrt2 = son_father[nrt2];
            }

            if (rt1 == rt2) continue;

            //else
            son_father[rt2] = rt1;
            for (auto &rt2_son : root_sons[rt2]) {
                root_sons[rt1].emplace(rt2_son);
                son_father[rt2_son] = rt1;
            }
            root_sons.erase(rt2);
        }
        return stones.size() - root_sons.size();
    }
};
int main() {

    //system("pause");
    return 0;
}
